[Medium] Prevent concurrent writes to VOLTA_HOME

charlespierce opened the pull request:
Closes #612 

Info
-----
* As described in #612, we currently have issues when multiple Volta processes try to mutate the VOLTA_HOME directory simultaneously.
* The most common case we've seen so far for this is when two processes try to download the same version of Node / Yarn at the same time (e.g. in an automated, parallel test).
* This PR alleviates those issues, based on the idea of double-checked locking:
    * First, we check if we need to download the tool.
    * If we do, then we acquire an exclusive file lock to a specific file in the Volta directory.
    * Once the lock is acquired, other processes may have already downloaded the tool we are looking at, so we check _again_.
    * If we detect that we _still_ need to download the tool / version, then because we have the lock, we can be confident that no other process is doing the same concurrently.
    * If, instead, the second check reveals that we don't need to download again, then another process must have taken care of it, so we can proceed.

Changes
-----
* Added the above-described double-checked locking behavior to fetching for all tools: Node, npm, Yarn, and Packages.
* Added similar behavior for the migrations, locking within `volta-migrate` before running detection of which migrations are needed.
* In all cases, errors creating or acquiring the lock fall back to the existing, non-concurrent behavior (e.g. immediately executing the fetch / migration

Tested
-----
* Added smoke tests that use an extra thread to run two concurrent installs to ensure that the changes work as expected (these tests fail without the changes).

Notes
-----
* For more context on double-checked locking, see the description in the [comments of the `double_checked_cell` crate](https://github.com/niklasf/double-checked-cell/blob/5907b25a493b48ec1516ee123b793cfc7be2272b/src/lib.rs#L264)

rwjblue commented:
Looks like this is building upon https://github.com/volta-cli/volta/pull/683, just calling out since it makes reviewing easier if you review commit by commit.

charlespierce commented:
@rwjblue Just updated the description with info about the approach used (and added the note about it building on #683)

charlespierce commented on crates/volta-core/src/tool/node/mod.rs:
```diff
@@ -104,17 +105,39 @@ impl Node {
+            // Additionally, if there is an error creating the lock or locking it, we act synchronously
+            let lock_name = format!("volta-fetch-node-{}", self.version.to_string());
+            if let Ok(lock) = NamedLock::create(&lock_name) {
+                if let Ok(_guard) = lock.lock() {
```
Definitely! That's a good call-out, we should have a spinner and a message about why we are waiting.

charlespierce commented:
Updated to abstract around creating the lock (and add a spinner for when we need to wait for the lock) and the double-checking logic.

Also switched to using file locking from the `fs2` crate, which is much more lightweight and mature than the `named_lock` crate.

I think this is in a good state now, so dropping the `[Experimental]`
